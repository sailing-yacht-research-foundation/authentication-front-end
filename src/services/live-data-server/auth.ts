import axios from "axios";
import { SYRF_SERVER } from "services/service-constants";
import syrfRequest from "utils/syrf-request";
import { v4 as uuidv4 } from 'uuid';

let uuid = localStorage.getItem('uuid');

if (uuid === null) {
    uuid = uuidv4();
    if (uuid)
        localStorage.setItem('uuid', uuid);
}

export const login = ({ email, password }) => {
    return axios.post(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/auth/login`, {
        username: email,
        password: password,
        devToken: SYRF_SERVER.DEV_TOKEN
    }).then(response => {
        return {
            success: true,
            token: response.data.token,
            user: response.data
        }
    }).catch(error => {
        return {
            success: false,
            error: error
        }
    })
}

export const anonymousLogin = () => {
    return axios.post(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/auth/anonymous-login`, {
        id: uuid,
        devToken: SYRF_SERVER.DEV_TOKEN
    }).then(response => {
        return {
            success: true,
            token: response.data?.token,
            data: response.data
        }
    }).catch(error => {
        return {
            success: false,
            error: error
        }
    })
}

export const validateToken = (token) => {
    return syrfRequest.post(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/auth/validate-token`, {
        token: token,
        renew: false
    }).then(response => {
        return {
            success: true,
            token: response.data.token,
            user: response.data
        }
    }).catch(error => {
        return {
            success: false,
            error: error
        }
    })
}


export const renewToken = (refreshToken) => {
    return syrfRequest.post(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/auth/renew-token`, {
        refresh_token: refreshToken
    })
    .then(response => {
        return {
            success: true,
            data: response.data
        }
    }).catch(error => {
        return {
            sucess: false,
            error: error
        }
    })
}

export const register = (data) => {
    return axios.post(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/users`, data)
    .then(response => {
        return {
            success: true,
            user: response.data.user
        }
    }).catch(error => {
        return {
            sucess: false,
            error: error
        }
    })
}

export const verifyEmail = () => {
    return syrfRequest.put(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/users/send-verify-email`)
    .then(response => {
        return {
            success: true,
        }
    }).catch(error => {
        return {
            sucess: false,
            error: error
        }
    })
}

export const sendForgotPassword = (email) => {
    return syrfRequest.put(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/users/forget-password`, {
        email: email
    })
    .then(response => {
        return {
            success: true,
        }
    }).catch(error => {
        return {
            sucess: false,
            error: error
        }
    })
}