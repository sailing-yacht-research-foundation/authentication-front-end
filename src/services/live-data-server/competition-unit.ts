import { SYRF_SERVER } from "services/service-constants";
import syrfService from 'utils/syrf-request';

export const getCompetitionUnitList = () => {
  return syrfService.get(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/competition-units`, {
  }).then(response => {
      return {
        success: true,
        data: response.data
      }
  }).catch(error => {
    return {
      success: false,
      error: error
    }
  });
}


export const getCompetitionUnitById = (id) => {
  return syrfService.get(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/competition-units/${id}`, {
  }).then(response => {
      return {
        success: true,
        data: response.data
      }
  }).catch(error => {
    return {
      success: false,
      error: error
    }
  });
};