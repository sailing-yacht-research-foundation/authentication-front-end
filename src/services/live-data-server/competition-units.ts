import moment from 'moment';
import { SYRF_SERVER } from 'services/service-constants';
import { formatServicePromiseResponse, parseKeyword } from 'utils/helpers';
import syrfRequest from 'utils/syrf-request';

export const search = (params) => {
    const query: any = {
        bool: {
            must: []
        }
    };

    query.bool.must.push({
        query_string: {
            query: parseKeyword(params.keyword),
        },

    });

    if (params.hasOwnProperty('from_date')
        && params.from_date !== ''
        && moment(params.from_date).isValid()) {
        query.bool.must.push({
            range: {
                "approx_start_time_ms": {
                    "gte": moment(params.from_date).set({ hour: 0, minute: 0, second: 0 }).unix() * 1000,
                }
            }
        });
    }

    if (params.hasOwnProperty('to_date')
        && params.to_date !== ''
        && moment(params.to_date).isValid()) {
        query.bool.must.push({
            range: {
                "approx_start_time_ms": {
                    "lt": moment(params.to_date).set({ hour: 23, minute: 59, second: 59 }).unix() * 1000,
                }
            }
        });
    }

    const searchParams: any = {
        query: query,
        sort: [
            "_score",
            {
                "approx_start_time_ms": {
                    "order": "desc"
                }
            },
        ]
    };

    searchParams._source = [
        "id", "source", "name", "approx_start_point", "start_country", "start_city", "start_year", "start_month",
        "approx_start_time_ms", "event_name", "event", "event_description", "isOpen", "allowRegistration", "status"
    ]; // only the fields we need
    searchParams.from = params.hasOwnProperty('page') ? ((Number(params.page) - 1) * Number(params?.size)) : 0;
    searchParams.size = params.size ?? 10;

    return formatServicePromiseResponse(syrfRequest.post(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/competition-units/search`, searchParams))
}

export const create = (calendarEventId, data) => {
    return formatServicePromiseResponse(syrfRequest.post(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/calendar-events/${calendarEventId}/competition-units`, {
        ...data
    }))
}

export const getAllByCalendarEventId = (calendarEventId, page, size = 10) => {
    return formatServicePromiseResponse(syrfRequest.get(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/calendar-events/${calendarEventId}/competition-units`, {
        params: {
            page: page,
            size: size,
        }
    }));
}

export const getAllCompetitionUnits = (page) => {
    const userId: any = localStorage.getItem('user_id');
    return formatServicePromiseResponse(syrfRequest.get(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/competition-units${!!userId ? `?createdById_eq=${userId}` : ''}`, {
        params: {
            page: page
        }
    }))
}

export const get = (calendarEventId, competitionUnitId) => {
    return formatServicePromiseResponse(syrfRequest.get(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/calendar-events/${calendarEventId}/competition-units/${competitionUnitId}`))
}

export const update = (calendarEventId, competitionUnitId, data) => {
    return formatServicePromiseResponse(syrfRequest.put(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/calendar-events/${calendarEventId}/competition-units/${competitionUnitId}`, {
        ...data
    }))
}

export const deleteCompetitionUnit = (calendarEventId, competitionUnitId) => {
    return formatServicePromiseResponse(syrfRequest.delete(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/calendar-events/${calendarEventId}/competition-units/${competitionUnitId}`))
}

export const getCompetitionUnitList = () => {
    return formatServicePromiseResponse(syrfRequest.get(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/competition-units`))
}

export const getCompetitionUnitById = (id) => {
    return formatServicePromiseResponse(syrfRequest.get(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/competition-units/${id}`));
};

export const searchScrapedRaceById = (id: string) => {
    const searchParams: any = {
        query: {
            match: {
                '_id': id
            }
        },
    };

    searchParams._source = ["id", "name", "approx_start_point", "start_country", "approx_start_time_ms", "url", "source", "isOpen", "allowRegistration"];

    return formatServicePromiseResponse(syrfRequest.post(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/competition-units/search`, searchParams))
}

export const cloneCourse = (fromCompetitionUnitId, toCompetitionUnitId, newName) => {
    return formatServicePromiseResponse(syrfRequest.put(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/competition-units/${toCompetitionUnitId}/clone-course`, {
        cloneFromCompetitionId: fromCompetitionUnitId,
        newName: newName
    }))
}

export const getAllCompetitionUnitsByEventIdWithSort = (calendarEventId, page) => {
    const userId: any = localStorage.getItem('user_id');
    return formatServicePromiseResponse(syrfRequest.get(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/calendar-events/${calendarEventId}/competition-units?sort=createdAt&srdir=-1${!!userId ? `&createdById_eq=${userId}` : ''}`, {
        params: {
            page: page
        }
    }))
}

export const getSimplifiedTracksByCompetitionUnit = (id: string) => {
    return formatServicePromiseResponse(syrfRequest.get(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/competition-units/${id}/simplified-tracks`))
}

export const getLegsByCompetitionUnit = (id: string) => {
    return formatServicePromiseResponse(syrfRequest.get(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/competition-units/${id}/legs`))
}

export const getTimeByCompetitionUnit = (id: string) => {
    return formatServicePromiseResponse(syrfRequest.get(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/competition-units/${id}/time`))
}

export const getCourseByCompetitionUnit = (id: string) => {
    return formatServicePromiseResponse(syrfRequest.get(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/competition-units/${id}/course`))
}

export const getSuggestion = (fieldName, word) => {
    const searchParams = {
        "suggest": {
            "autocomplete": {
                "prefix": word,
                "completion": {
                    "field": `${fieldName}.completion`,
                    "skip_duplicates": true,
                    "fuzzy": {
                        "fuzziness": 1
                    }
                }
            }
        }
    };

    return formatServicePromiseResponse(syrfRequest.post(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/competition-units/search`, searchParams))
}

export const getRaceViewsCount = (competitionUnitId) => {
    return formatServicePromiseResponse(syrfRequest.get(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/competition-units/${competitionUnitId}/viewers-count`));
}

export const stopRace = (competitionUnitId) => {
    return formatServicePromiseResponse(syrfRequest.put(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/competition-units/${competitionUnitId}/stop-race`));
}

export const startRace = (competitionUnitId) => {
    return formatServicePromiseResponse(syrfRequest.put(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/competition-units/${competitionUnitId}/prepare-race`));
}

/**
 * Get live and happening soon races.
 * @param duration // in months
 * @param distance in mile
 * @param page 
 * @param size 
 * @param coordinate { lat, lon }
 * @returns 
 */
export const getLiveAndUpcomingRaces = (duration: number = 1, distance: number = 1000, page: number = 1, size: number = 10, coordinate) => {
    const query: any = {
        bool: {
            must: {},
            filter: [],
            should: [],
        },
    };

    if (coordinate) {
        query.bool.filter.push({
            "geo_distance": {
                "distance": `${distance}nmi`,
                "approx_start_point.coordinates": {
                    "lat": coordinate.lat,
                    "lon": coordinate.lon
                }
            }
        })
    }

    query.bool.must.bool = { // endtime is greater than now or starttime greater than now AND less than duration.
        should: [
            {
                "range": {
                    "approximateEndTime": {
                        "gte": "now"
                    }
                }
            },
            {
                "range": {
                    "approx_start_time_ms": {
                        "gte": moment().set({ hour: 0, minute: 0, second: 0 }).unix() * 1000,
                        "lt": moment().add(duration, "months").set({ hour: 23, minute: 59, second: 59 }).unix() * 1000,
                    }
                }
            }
        ]
    }

    const searchParams: any = {
        query: query,
        sort: [
            {
                "approx_start_time_ms": {
                    "order": "asc"
                }
            },
        ]
    };

    if (coordinate) {
        searchParams.sort.push({
            "_geo_distance": {
                "approx_start_point.coordinates": {
                    "lat": coordinate.lat,
                    "lon": coordinate.lon
                },
                "order": "asc",
            }
        })
    }

    searchParams._source = [
        "id", "source", "name", "approx_start_point", "start_country", "start_city", "start_year", "start_month",
        "approx_start_time_ms", "event_name", "event", "event_description", "isOpen", "allowRegistration", "status"
    ]; // only the fields we need
    searchParams.from = (page - 1) * size;
    searchParams.size = size;

    return formatServicePromiseResponse(syrfRequest.post(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/competition-units/search`, searchParams))
}