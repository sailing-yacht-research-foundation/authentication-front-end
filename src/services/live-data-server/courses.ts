import { SYRF_SERVER } from 'services/service-constants';
import syrfService from 'utils/syrf-request';

export const list = () => {
    return syrfService.get(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/courses`).then(response => {
        return {
            success: true,
            data: response.data
        }
    }).catch(error => {
        return {
            success: false,
            error: error
        }
    });
}

export const create = (competitionUnitId, courseSequencedGeometries) => {
    return syrfService.post(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/courses`, {
        courseSequencedGeometries: courseSequencedGeometries,
        competitionUnitId: competitionUnitId
    }).then(response => {
        return {
            success: true,
            data: response.data
        }
    }).catch(error => {
        return {
            success: false,
            error: error
        }
    });
}

export const deleteCourse = (courseId) => {
    return syrfService.delete(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/courses/${courseId}`).then(response => {
        return {
            success: true,
            data: response.data
        }
    }).catch(error => {
        return {
            success: false,
            error: error
        }
    });
}

export const update = (competitionUnitId, courseId, courseSequencedGeometries) => {
    return syrfService.put(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/courses/${courseId}`, {
        courseSequencedGeometries: courseSequencedGeometries,
        competitionUnitId: competitionUnitId
    }).then(response => {
        return {
            success: true,
            data: response.data
        }
    }).catch(error => {
        return {
            success: false,
            error: error
        }
    });
}

export const updateCourseGeometry = (courseId, courseSequencedGeometries) => {
    console.log(courseSequencedGeometries);
    return syrfService.put(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/courses/${courseId}/sequenced`, courseSequencedGeometries).then(response => {
        return {
            success: true,
            data: response.data
        }
    }).catch(error => {
        return {
            success: false,
            error: error
        }
    });
}

export const getByCompetitionUnit = (competitionUnitId) => {
    return syrfService.get(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/competition-units/${competitionUnitId}/course`).then(response => {
        return {
            success: true,
            data: response.data
        }
    }).catch(error => {
        return {
            success: false,
            error: error
        }
    });
}

export const getById = (courseId) => {
    return syrfService.get(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/courses/${courseId}`)
        .then(response => {
            return {
                success: true,
                data: response.data
            }
        }).catch(error => {
            return {
                success: false,
                error: error
            }
        });
}