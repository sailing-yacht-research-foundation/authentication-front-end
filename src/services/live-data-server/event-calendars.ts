import { SYRF_SERVER } from 'services/service-constants';
import { CalendarEvent } from 'types/CalendarEvent';
import { EventState } from 'utils/constants';
import { formatServicePromiseResponse } from 'utils/helpers';
import syrfRequest from 'utils/syrf-request';

export const getAll = () => {
    const userId: any = localStorage.getItem('user_id');
    return formatServicePromiseResponse(syrfRequest.get(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/calendar-events${!!userId ? `?createdById_eq=${userId}` : ''}`))
}

export const getMany = (page: string, size: number = 10) => {
    return formatServicePromiseResponse(syrfRequest.get(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/calendar-events/my-events?isPrivate_eq=false`, {
        params: {
            page,
            size
        }
    }))
}

export const get = (id: string) => {
    return formatServicePromiseResponse(syrfRequest.get(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/calendar-events/${id}`))
}

export const create = (data) => {
    return formatServicePromiseResponse(syrfRequest.post(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/calendar-events`, {
        ...data
    }))
}

export const update = (id: string, data) => {
    return formatServicePromiseResponse(syrfRequest.put(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/calendar-events/${id}`, {
        ...data
    }));
}

export const deleteEvent = (id: string) => {
    return formatServicePromiseResponse(syrfRequest.delete(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/calendar-events/${id}`))
}

export const downloadIcalendarFile = (event: Partial<CalendarEvent>) => {
    return syrfRequest.get(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/calendar-events/${event.id}/ics`, { responseType: 'blob' }).then(response => {
        const url = window.URL.createObjectURL(new Blob([response.data]));
        const link = document.createElement('a');
        link.href = url;
        link.setAttribute('download', `${event.name}.ics`); //or any other extension
        document.body.appendChild(link);
        link.click();
    }).catch(error => {
        return {
            success: false,
            error: error
        }
    })
}

export const getEditors = (eventId: string) => {
    return formatServicePromiseResponse(syrfRequest.get(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/calendar-events/${eventId}/editors`))
}

export const addEditor = (eventId: string, editorId: string) => {
    return formatServicePromiseResponse(syrfRequest.put(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/calendar-events/${eventId}/add-editors`, {
        userIds: [
            editorId
        ]
    }))
}

export const removeEditor = (eventId: string, editorId: string) => {
    return formatServicePromiseResponse(syrfRequest.delete(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/calendar-events/${eventId}/remove-editors`, {
        data: {
            userIds: [
                editorId
            ]
        }
    }));
}

export const scheduleCalendarEvent = (eventId: string) => {
    return formatServicePromiseResponse(syrfRequest.patch(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/calendar-events/${eventId}/status`, {
        status: EventState.SCHEDULED
    }));
}

export const toggleOpenForRegistration = (eventId: string, allowRegistration: boolean) => {
    return formatServicePromiseResponse(syrfRequest.patch(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/calendar-events/${eventId}/manage-registration`, {
        allowRegistration
    }));
}

export const closeCalendarEvent = (eventId: string) => {
    return formatServicePromiseResponse(syrfRequest.put(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/calendar-events/${eventId}/stop`))
}

export const cancelCalendarEvent = (eventId: string) => {
    return formatServicePromiseResponse(syrfRequest.patch(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/calendar-events/${eventId}/status`, {
        status: EventState.CANCELED
    }));
}

export const stopEvent = (calendarEventId: string) => {
    return formatServicePromiseResponse(syrfRequest.put(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/calendar-events/${calendarEventId}/stop`));
}

export const getEventRegisteredVessels = (calendarEventId: string, page: number) => {
    return formatServicePromiseResponse(syrfRequest.get(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/calendar-events/${calendarEventId}/registered-vessels`, {
        params: {
            page
        }
    }));
}

export const uploadPdfs = (calendarEventId: string, formData: FormData) => {
    return formatServicePromiseResponse(syrfRequest.put(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/calendar-events/${calendarEventId}/upload-pdfs`, formData,  {
        headers: { "content-type": "multipart/form-data" }
    }));
}

export const sendMessageToVesselParticipants = (calendarEventId: string, data) => {
    return formatServicePromiseResponse(syrfRequest.put(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/calendar-events/${calendarEventId}/send-messages`, data));
}

export const payForEvent = (calendarEventId: string) => {
    return formatServicePromiseResponse(syrfRequest.post(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/calendar-events/${calendarEventId}/pay`));
}

export const getDetailedEventParticipantsInfo = (calendarEventId: string) => {
    return formatServicePromiseResponse(syrfRequest.get(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/calendar-events/${calendarEventId}/participants/complete-data`));
}

export const agreeToWaiver = (calendarEventId: string, waiverType: string) => {
    return formatServicePromiseResponse(syrfRequest.post(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/calendar-events/${calendarEventId}/agree-waiver`, {
        waiverType
    }));
}

export const getDetailedEventParticipantInfoById = (calendarEventId: string, participantId: string) => {
    return formatServicePromiseResponse(syrfRequest.get(`${SYRF_SERVER.API_URL}${SYRF_SERVER.API_VERSION}/calendar-events/${calendarEventId}/participants/${participantId}/complete-data`));
}