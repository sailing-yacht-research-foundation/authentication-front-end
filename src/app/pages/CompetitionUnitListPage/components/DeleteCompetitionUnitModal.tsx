import React from 'react';
import { Modal } from 'antd';
import { toast } from 'react-toastify';
import styled from 'styled-components';
import { useTranslation } from 'react-i18next';
import { deleteCompetitionUnit } from 'services/live-data-server/competition-units';
import { translations } from 'locales/translations';
import { showToastMessageOnRequestError } from 'utils/helpers';
import { CompetitionUnit } from 'types/CompetitionUnit';

interface IDeleteCompetitionUnitModal {
    competitionUnit: Partial<CompetitionUnit>,
    showDeleteModal: boolean,
    setShowDeleteModal: Function,
    onCompetitionUnitDeleted: Function
}

export const DeleteCompetitionUnitModal = (props: IDeleteCompetitionUnitModal) => {

    const { t } = useTranslation();

    const [isLoading, setIsLoading] = React.useState<boolean>(false);

    const {
        competitionUnit,
        showDeleteModal,
        setShowDeleteModal,
        onCompetitionUnitDeleted
    } = props;

    const performDeleteCompetitionUnit = async () => {
        setIsLoading(true);
        const response = await deleteCompetitionUnit(competitionUnit.calendarEventId, competitionUnit.id);
        setIsLoading(false);

        setShowDeleteModal(false);

        if (response.success) {
            toast.success(t(translations.delete_competition_unit_modal.successfully_deleted, { name: competitionUnit.name }));
            onCompetitionUnitDeleted();
        } else {
            showToastMessageOnRequestError(response.error);
        }
    }

    return (
        <Modal
            confirmLoading={isLoading}
            title={t(translations.delete_competition_unit_modal.are_you_sure_you_want_to_delete)}
            visible={showDeleteModal}
            onOk={performDeleteCompetitionUnit}
            onCancel={() => {
                setShowDeleteModal(false);
            }}>
            <ModalMessage>{t(translations.delete_competition_unit_modal.you_will_delete)}</ModalMessage>
        </Modal>
    )
}

const ModalMessage = styled.div`
    margin: 0 5px;
    margin-bottom: 15px;
`;