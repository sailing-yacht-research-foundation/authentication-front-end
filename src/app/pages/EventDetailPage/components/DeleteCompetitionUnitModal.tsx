import React from 'react';
import { Modal } from 'antd';
import { toast } from 'react-toastify';
import styled from 'styled-components';
import { useTranslation } from 'react-i18next';
import { deleteCompetitionUnit } from 'services/live-data-server/competition-units';
import { translations } from 'locales/translations';
import { showToastMessageOnRequestError } from 'utils/helpers';

export const DeleteCompetitionUnitModal = (props) => {

    const { t } = useTranslation();

    const {
        competitionUnit,
        showDeleteModal,
        setShowDeleteModal,
        onCompetitionUnitDeleted
    } = props;

    const performDeleteCompetitionUnit = async () => {
        const response = await deleteCompetitionUnit(competitionUnit.calendarEventId, competitionUnit.id);

        setShowDeleteModal(false);

        if (response.success) {
            toast.success(t(translations.delete_competition_unit_modal.successfully_deleted, { name: competitionUnit.name }));
            onCompetitionUnitDeleted();
        } else {
            showToastMessageOnRequestError(response.error);
        }
    }

    return (
        <Modal
            title={t(translations.delete_competition_unit_modal.are_you_sure_you_want_to_delete)}
            visible={showDeleteModal}
            onOk={() => {
                performDeleteCompetitionUnit();
            }}
            onCancel={() => {
                setShowDeleteModal(false);
            }}>
            <ModalMessage>{t(translations.delete_competition_unit_modal.you_will_delete)}</ModalMessage>
        </Modal>
    )
}

const ModalMessage = styled.div`
    margin: 0 5px;
    margin-bottom: 15px;
`;