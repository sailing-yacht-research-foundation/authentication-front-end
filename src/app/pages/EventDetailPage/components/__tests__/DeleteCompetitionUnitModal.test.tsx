import * as React from 'react';
import { DeleteCompetitionUnitModal } from '../DeleteCompetitionUnitModal';
import * as DeleteCompetitionUnitModalModule from '../DeleteCompetitionUnitModal';
import { createRenderer } from 'react-test-renderer/shallow';
import { i18n } from 'locales/i18n';
import { translations } from 'locales/translations';
import { act, fireEvent, render, } from '@testing-library/react';
import * as CompetitionUnitModule from 'services/live-data-server/competition-units';
import * as Helpers from 'utils/helpers';

const shallowRenderer = createRenderer();

const modalShowComponent = (
    <DeleteCompetitionUnitModal competitionUnit={{}}
        showDeleteModal={true}
        setShowDeleteModal={jest.fn}
        onCompetitionUnitDeleted={jest.fn} />
);

describe('DeleteCompetitionUnitModal', () => {
    it('should render and match snapshot', () => {
        shallowRenderer.render(modalShowComponent);
        const renderedOutput = shallowRenderer.getRenderOutput();
        expect(renderedOutput).toMatchSnapshot();
    });

    // Use 'it' to test a single attribute of a target
    it('Hide if show modal flag is false', () => {

        shallowRenderer.render(<DeleteCompetitionUnitModal competitionUnit={{}}
            showDeleteModal={false}
            setShowDeleteModal={jest.fn}
            onCompetitionUnitDeleted={jest.fn} />);

        const renderedOutput = shallowRenderer.getRenderOutput();
        expect(renderedOutput.props.visible).not.toBeTruthy();
    });

    it('Show if show modal flag is true', async () => {
        shallowRenderer.render(modalShowComponent);

        const renderedOutput = shallowRenderer.getRenderOutput();
        expect(renderedOutput.props.visible).toBeTruthy();
    });

    it('Modal title and content should match i18next translation language', async () => {
        const t = await i18n;

        shallowRenderer.render(modalShowComponent);

        expect(t(shallowRenderer.getRenderOutput().props.title)).toBe(t(translations.delete_competition_unit_modal.are_you_sure_you_want_to_delete));
        expect(t(shallowRenderer.getRenderOutput().props.children.props.children)).toBe(t(translations.delete_competition_unit_modal.you_will_delete));
    });

    it("It should have been rendered with correct received props", () => {
        const deleteCompetitionUnitSpy = jest.spyOn(DeleteCompetitionUnitModalModule, 'DeleteCompetitionUnitModal');

        render(<DeleteCompetitionUnitModal event={{}} showModal={true} setShowModal={jest.fn} reloadParent={jest.fn} />);

        expect(deleteCompetitionUnitSpy).toHaveBeenCalledWith({
            event: {},
            showModal: true,
            setShowModal: jest.fn,
            reloadParent: jest.fn,
        }, {});
    });

    it("It should call setShowModal when pressing cancel text", async () => {
        const setShowModalMock = jest.fn();

        const { getByText } = render(<DeleteCompetitionUnitModal competitionUnit={{}}
            showDeleteModal={true}
            setShowDeleteModal={setShowModalMock}
            onCompetitionUnitDeleted={jest.fn} />);

        act(() => {
            fireEvent.click(getByText('Cancel'));
        });

        expect(setShowModalMock).toHaveBeenCalledWith(false);
    });

    it("It should call deleteCompetitionUnit when pressing ok text", async () => {
        let competitionUnitModuleMock;
        const onCompetitionUnitDeletedMock = jest.fn();
        competitionUnitModuleMock =  jest.spyOn(CompetitionUnitModule, 'deleteCompetitionUnit').mockImplementation(() => {
            return Promise.resolve({
                success: true
            })
        }); 
        
        const { getByText } = render(<DeleteCompetitionUnitModal competitionUnit={{}}
            showDeleteModal={true}
            setShowDeleteModal={jest.fn}
            onCompetitionUnitDeleted={onCompetitionUnitDeletedMock} />);

        act(() => {
            fireEvent.click(getByText('OK'));
        });

        expect(competitionUnitModuleMock).toHaveBeenCalled();
        await act(() => Promise.resolve()); // wait for deleteCompetitionUnit call.
        expect(onCompetitionUnitDeletedMock).toHaveBeenCalled();
    });

    it("It should call showToastOnServerResponseError when pressing ok text with failed request", async () => {
        const competitionUnitModuleMock =  jest.spyOn(CompetitionUnitModule, 'deleteCompetitionUnit').mockImplementation(() => {
            return Promise.resolve({
                success: false
            })
        });
        const showToastMessageOnRequestErrorSpy = jest.spyOn(Helpers, 'showToastMessageOnRequestError'); 
        
        const { getByText } = render(<DeleteCompetitionUnitModal competitionUnit={{}}
            showDeleteModal={true}
            setShowDeleteModal={jest.fn}
            onCompetitionUnitDeleted={jest.fn} />);

        act(() => {
            fireEvent.click(getByText('OK'));
        });

        expect(competitionUnitModuleMock).toHaveBeenCalled();
        await act(() => Promise.resolve()); // wait for deleteCompetitionUnit call.
        expect(showToastMessageOnRequestErrorSpy).toHaveBeenCalled();
    });
});