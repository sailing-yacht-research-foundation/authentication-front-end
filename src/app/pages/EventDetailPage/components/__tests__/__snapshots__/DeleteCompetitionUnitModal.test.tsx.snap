// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeleteCompetitionUnitModal should render and match snapshot 1`] = `
<Modal
  confirmLoading={false}
  okType="primary"
  onCancel={[Function]}
  onOk={[Function]}
  title="Are you sure you want to delete this race?"
  visible={true}
  width={520}
>
  <styled.div>
    You will delete this race along with all associated information, are you sure you want to continue?
  </styled.div>
</Modal>
`;
